source/array_list.c

```diff
@@ -260,3 +260,34 @@ int aws_array_list_set_at(struct aws_array_list *list, const void *val, size_t i
     return AWS_OP_SUCCESS;
 }
 
+void aws_array_list_swap(struct aws_array_list *list, size_t a, size_t b) {
+    assert(a < list->length);
+    assert(b < list->length);
+
+    void *item1, *item2;
+    aws_array_list_get_at_ptr(list, &item1, a);
+    aws_array_list_get_at_ptr(list, &item2, b);
+
+    /* copy bytes sizeof(size_t) at a time */
+    size_t sb_pointer_size_count = list->item_size / sizeof(size_t);
+    uint8_t temp[sizeof(size_t)];
+    for(size_t i = 0; i < sb_pointer_size_count; i++) {
+        memcpy(temp, item1, sizeof(size_t));
```

Comment:
As a performance note: The compiler has to assume that these items may overlap arbitrarily, which prevents it from doing optimizations like autovectorization or loop unrolling.

I'd suggest you 1) bail out if `a == b`, and then delegate to a helper `void aws_memswap(void * restrict a, void * restrict b, size_t len)`, which will tell the compiler that these two objects cannot overlap.